test/utils.test.js

```diff
@@ -195,14 +195,52 @@ describe('processInputs', () => {
     const res = utils.processInputs({ a: 'string', one: 'number', an: 'integer' }, { })
     expect(res).toEqual({ a: '', one: 0, an: 0 })
   })
-  // eslint-disable-next-line no-template-curly-in-string
   test('input = { an: $undefEnvVar, a: $definedEnvVar, another: $definedEnvVar, the: ${definedEnvVar}, one: ${ definedEnvVar  } }, params = { a: 123 }', () => {
     process.env.definedEnvVar = 'giraffe'
-    // eslint-disable-next-line no-template-curly-in-string
     const res = utils.processInputs({ an: '$undefEnvVar', a: '$definedEnvVar', another: '$definedEnvVar', the: '${definedEnvVar}', one: '${ definedEnvVar  }' }, { a: 123 })
     expect(res).toEqual({ a: 123, another: 'giraffe', an: '', the: 'giraffe', one: 'giraffe' })
     delete process.env.definedEnvVar
   })
+  test('invalid input returns undefined (coverage)', () => {
+    let res = utils.processInputs({}, { a: 123 })
+    expect(res).toEqual({})
+    res = utils.processInputs(undefined, { a: 123 })
+    expect(res).toEqual(undefined)
+    res = utils.processInputs('string', { a: 123 })
+    expect(res).toEqual(undefined)
+  })
+  test('nested inputs with default and params, {foo: ${bar}, config: {nestedFoo: {bar}}', () => {
+    process.env.BAR = 'itWorks'
+    process.env.BAR_VAR = 'barVar'
+    process.env.FOO = 'fooo'
+    const input = {
+      stuff: '$BAR_VAR $BAR_VAR, ${ BAR_VAR }, $FOO',
+      foo: '${BAR}',
+      bar: {
+        default: '${BAR}, $BAR, ${FOO}'
+      },
+      config: {
+        nestedFoo: {
+          extraNested: '${BAR}, $BAR, ${FOO}'
+        },
+        a: 'I will be replaced'
+      }
+    }
+    const expectedOutput = {
+      stuff: 'barVar barVar, barVar, fooo',
+      foo: process.env.BAR,
+      bar: `${process.env.BAR}, ${process.env.BAR}, ${process.env.FOO}`,
+      config: {
+        nestedFoo: {
+          extraNested: `${process.env.BAR}, ${process.env.BAR}, ${process.env.FOO}`
+        },
+        a: 123
```

Comment:
why would `a` be replaced by the params `{ a: 123 }` since `a` is nested under `config.a` ?

Reply:
Right, I think I understand why. 
Params are supposed to replace env values so, it should do it only for the front keys and it shouldn't go deeper.

Reply:
@shazron @purplecabbage 
```
test('input = { I : { value: am } }, params = { I : { value: nitpicking } }', () => {
    // note: is this relevant?
    const res = utils.processInputs({ I: { value: 'am' } }, { I: { value: 'nitpicking' } })
    expect(res).toEqual({ I: { value: 'nitpicking' } })
  })
```
  
  Though, what about this one? Is this relevant? 
  I have eliminated it since I feel it's not relevant, but I could use a confirmation here.

Reply:
Right - the params are for top level deep properties.

As for that piece of code, it doesn't really replace `I.value` it replaces everything in `I` so the test should reflect that (I don't think that is clear)